language-examples/remedial-js/8.0-JS-101/README.md

```diff
@@ -0,0 +1,307 @@
+
+
+//////////////////////////
+// WITHOUT METHOD CHAINING
+
+var $div = $('#my-div');         // assign to var
+
+$div.css('background', 'blue');  // set BG
+$div.height(100);                // set height
+$div.fadeIn(200);                // show element
+
+
+///////////////////////
+// WITH METHOD CHAINING
+
+$('#my-div').css('background', 'blue').height(100).fadeIn(200);
+
+// often broken to multiple lines:
+$('#my-div')
+  .css('background', 'blue')
+  .height(100)
+  .fadeIn(200);
+```
+
+is really equivalent to this:
+
+```javascript
+  angular.module('ngcourse').controller(...);
+```
+
+# Exercise 1
+
+To support this `fluent` programming style, Lets define our own class and add method chaining.  Notice each prototype method is returning an object.
+This code is running live in the exercise.js file, so you can open the browser, navigate to this exercise, and see the code in action in the console.
+
+```javascript
+// define the class
+var Kitten = function() {
+  this.name = 'Garfield';
+  this.color = 'brown';
+  this.gender = 'male';
+};
+
+Kitten.prototype.setName = function(name) {
+  this.name = name;
+  return this;
+};
+
+Kitten.prototype.setColor = function(color) {
+  this.color = color;
+  return this;
+};
+
+Kitten.prototype.setGender = function(gender) {
+  this.gender = gender;
+  return this;
+};
+
+Kitten.prototype.save = function() {
+  console.log(
+    'saving ' + this.name + ', the ' +
+    this.color + ' ' + this.gender + ' kitten...'
+  );
+
+  // save to database here...
+
+  return this;
+};
+
+// Now use our new class with method chaining
+var bob = new Kitten();
+
+var tmp = bob.setName('Bob');
+tmp.setColor('black');
+
+console.log(tmp === bob);
+
+// OUTPUT:
+// > true
+
+// WE CAN EVEN MAKE IT SHORTER:
+
+new Kitten()
+  .setName('Bob')
+  .setColor('black')
+  .setGender('male')
+  .save();
+
+// OUTPUT:
+// > saving Bob, the black male kitten...
+```
+
+# JavaScript Functions
+
+In languages such as Java, arguments to functions and methods can be objects
+or primitives, but not functions. In JavaScript, however, functions are first
+class citizens. A function can be passed into another function as a parameter.
+A function can return a function. Functions can also be assigned to variables.
+
+JavaScript allows two ways of defining a function. In the first method, called
+"function declaration", a new function is defined in the current
+scope and is given a name:
+
+```javascript
+  function foo() {
+    // do something
+  }
+```
+
+We will be able to refer to this function as `foo` in the current scope. It is
+worth noting that functions defined using function declarations are "hoisted"
+in JavaScript. Regardless of where you define a function in the current scope,
+JavaScript would act as if the function was defined up front. So, this is
+perfectly valid:
+
+```javascript
+  // Call a function.
+  foo();
+
+  // Now provide a definition.
+  function foo() {
+    // do something
+  }
```

Comment:
I meant in this readme there is an example of hoisting. Beyond a one line explanation I'm not sure if there is much more to cover for hoisting?